mantidproject/mantid
# Implementation of Voxels in Mantid as GridDetector

**Description of work.**

The concept of a 3D detector bank has been added to mantid in the form of the GridDetector. This contains the concepts of layers of detectors within a bank in an attempt to solve the issue of visualising voxel based instruments.

**To test:**

In addition to the test below:

 - **Ensure other instruments which contain tubes, rectangular banks etc. are loaded and displayed correctly.**

Launch Mantid and run the script below. Load the instrument view for the workspace generated and select `ShowSingleLayer` use the slider to view separate bank layers.:

```python
import numpy as np

def histEventList(evlist):
    bins = np.histogram(evlist[:, 1], 500)
    binnedData=np.zeros([500,4*16*48])
    for i in range(4*16*48):
        ind = evlist[:, 0] == i
        tmp=evlist[ind,1]
       
        binnedEvents = np.histogram(tmp, bins[1])
        
        binnedData[:,i]=binnedEvents[0]

    return bins, binnedData

evlst=np.load([Your/Path/To/EventList.dat.npy])

Eventlist.dat.npy])

bins,binneddata=histEventList(evlst)
test_grid=CreateWorkspace(bins[1],np.transpose(binneddata),NSpec=4*16*48,UnitX='tof')
LoadInstrument("test_grid", True, Filename=[Your/Path/To/BasicGrid_Definition.xml]) 
```
Files here:
[TestFiles.zip](https://github.com/mantidproject/mantid/files/2400324/TestFiles.zip)

Fixes #23073. 




---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
